### PR TITLE
DO NOT MERGE: testing Agentic workflows #2

### PR DESCRIPTION
**Pull request by ruokun-niu**
# Description

_Please explain the changes you've made._

## Type of change

- This pull request fixes a bug in Drasi and has an approved issue (issue link required).
- This pull request adds or changes features of Drasi and has an approved issue (issue link required).
- This pull request is a minor refactor, code cleanup, test improvement, or other maintenance task and doesn't change the functionality of Drasi (issue link optional).

Fixes: #issue_number

### CODE REVIEW

**Comment by ruokun-niu** [comment]
## ✅ Correctness Review

### `core/src/models/timestamp_range.rs`

The `contains` logic is correct:
- Upper bound early-return uses strict `>`, so `ts == self.to` correctly passes through (inclusive upper bound).
- `Included` lower bound is inclusive (`>=`), `StartFromPrevious` is exclusive (`>`), matching the doc comment.
- Inverted ranges (`from > to`) silently return `false` for all inputs — this is safe and consistent, though undocumented.

🔵 **Nit** — `contains` borrows `self.from` to get a `&ElementTimestamp`, then immediately dereferences it. Since `ElementTimestamp` is `u64` (a `Copy` type), the borrow is unnecessary:

```rust
// Before
match &self.from {
    TimestampBound::Included(from) => ts >= *from,
    TimestampBound::StartFromPrevious(from) => ts > *from,
}

// After
match self.from {
    TimestampBound::Included(from) => ts >= from,
    TimestampBound::StartFromPrevious(from) => ts > from,
}
```

---

### `.github/workflows/pr-reviewers.yml`

The workflow logic is correct. The `$\{\{ steps.reviewers.outputs.workflows }}` expansion in the `for` loop relies on shell word-splitting to iterate over space-separated filenames — this works correctly for the `.yml` filenames used here. The `$GH_TOKEN` null-check is reachable (empty secret → empty string → `[ -z "" ]` is true) and correctly fails fast with a useful error message.

No correctness issues in the workflow file.



> [!WARNING]
> <details>
> <summary>Firewall blocked 1 domain</summary>
>
> The following domain was blocked by the firewall during workflow execution:
>
> - `drasi.io`
>> To allow these domains, add them to the `network.allowed` list in your workflow frontmatter:
>
> ```yaml
> network:
>   allowed:
>     - defaults
>     - "drasi.io"
> ```
>
> See [Network Configuration](https://github.github.com/gh-aw/reference/network/) for more information.
>
> </details>


> Generated by [pr-correctness-reviewer](https://github.com/drasi-project/.github/actions/runs/25402787054/agentic_workflow) · ● 149.2K · [◷](https://github.com/search?q=repo%3Adrasi-project%2Fdrasi-core+%22gh-aw-workflow-call-id%3A+drasi-project%2F.github%2Fpr-correctness-reviewer%22&type=issues)

**Comment by ruokun-niu** [comment]
## ✅ Correctness Review

Findings are in `core/src/sample_utils.rs`. `timestamp_range.rs` logic is correct.

---

### 🔴 Blocker — Off-by-one in `take_first` (`sample_utils.rs`, line 44)

`0..=n` iterates **n+1** times (indices 0 through n inclusive). If `src.len() == n`, `src[n]` is out of bounds and panics at runtime.

```rust
// Before
for i in 0..=n {
    out.push(src[i]);
}

// After
for i in 0..n {
    out.push(src[i]);
}
```

---

### 🔴 Blocker — Divide-by-zero in `average` (`sample_utils.rs`, line 31)

`values.len()` is 0 when the slice is empty, causing a panic on `total / values.len() as u64`.

```rust
// Before
pub fn average(values: &[u64]) -> u64 {
    let mut total: u64 = 0;
    for v in values { total += v; }
    total / values.len() as u64
}

// After
pub fn average(values: &[u64]) -> Option<u64> {
    if values.is_empty() {
        return None;
    }
    let total: u64 = values.iter().sum();
    Some(total / values.len() as u64)
}
```

---

### 🔴 Blocker — `unwrap()` on user input in `parse_ids` (`sample_utils.rs`, line 24)

Panics on any non-integer token. Non-test public APIs must not panic on bad input.

```rust
// Before
pub fn parse_ids(input: &str) -> Vec<u32> {
    let mut out = Vec::new();
    for part in input.split(',') {
        let n: u32 = part.trim().parse().unwrap();
        out.push(n);
    }
    out
}

// After
pub fn parse_ids(input: &str) -> Result<Vec<u32>, std::num::ParseIntError> {
    input.split(',').map(|p| p.trim().parse::<u32>()).collect()
}
```

---

### 🟡 Should-Fix — Silent lossy truncation in `report_count` (`sample_utils.rs`, line 73)

`total as u32` silently wraps values > `u32::MAX`. Either return `u64`, use `u32::try_from`, or document the truncation as intentional with a `#[allow(clippy::cast_possible_truncation)]`.

```rust
// Option A: return u64
pub fn report_count(&self, total: u64) -> u64 { total }

// Option B: checked conversion
pub fn report_count(&self, total: u64) -> u32 {
    u32::try_from(total).unwrap_or(u32::MAX)
}
```

---

### 🟡 Should-Fix — `unwrap()` on `Mutex::lock()` in `LastSeenCache` (`sample_utils.rs`, lines 60, 65)

A poisoned mutex causes a panic. In non-test library code, prefer explicit poison handling:

```rust
pub fn record(&self, key: &str, value: i64) {
    let mut guard = self.inner.lock().unwrap_or_else(|e| e.into_inner());
    guard.insert(key.to_string(), value);
}
```

---

### 🟡 Should-Fix — u64 overflow risk in `average` (`sample_utils.rs`, line 33)

Summing many large `u64` values can overflow in debug builds (panic) and wrap silently in release builds. Use `u128` accumulator or `checked_add`:

```rust
let total: u128 = values.iter().map(|&v| v as u128).sum();
Some((total / values.len() as u128) as u64)
```



> [!WARNING]
> <details>
> <summary>Firewall blocked 1 domain</summary>
>
> The following domain was blocked by the firewall during workflow execution:
>
> - `drasi.io`
>> To allow these domains, add them to the `network.allowed` list in your workflow frontmatter:
>
> ```yaml
> network:
>   allowed:
>     - defaults
>     - "drasi.io"
> ```
>
> See [Network Configuration](https://github.github.com/gh-aw/reference/network/) for more information.
>
> </details>


> Generated by [pr-correctness-reviewer](https://github.com/drasi-project/.github/actions/runs/25403315734/agentic_workflow) · ● 120.1K · [◷](https://github.com/search?q=repo%3Adrasi-project%2Fdrasi-core+%22gh-aw-workflow-call-id%3A+drasi-project%2F.github%2Fpr-correctness-reviewer%22&type=issues)

**Comment by ruokun-niu** [comment]
## ✅ Correctness Review

Reviewed files: `core/src/sample_utils.rs`, `core/src/models/timestamp_range.rs`, `core/src/lib.rs`, `.github/workflows/pr-reviewers.yml`.

---

### `core/src/sample_utils.rs`

🔴 **Blocker — off-by-one in `take_first` (line 44)**

`0..=n` is inclusive, iterating `n+1` times and accessing `src[n]`, which panics when `src` has exactly `n` elements.

```rust
// Before
for i in 0..=n {
    out.push(src[i]);
}

// After
for i in 0..n {
    out.push(src[i]);
}
```

---

🔴 **Blocker — division by zero in `average` (line 34)**

When `values` is empty, `values.len()` is `0` and integer division panics.

```rust
// Before
pub fn average(values: &[u64]) -> u64 {
    let mut total: u64 = 0;
    for v in values {
        total += v;
    }
    total / values.len() as u64
}

// After
pub fn average(values: &[u64]) -> Option<u64> {
    if values.is_empty() {
        return None;
    }
    let total: u64 = values.iter().sum();
    Some(total / values.len() as u64)
}
```

---

🟡 **Should-Fix — `unwrap()` on user-controlled input in `parse_ids` (line 24)**

`part.trim().parse().unwrap()` panics on any non-integer token. Non-test code should propagate the error.

```rust
// Before
pub fn parse_ids(input: &str) -> Vec<u32> {
    let mut out = Vec::new();
    for part in input.split(',') {
        let n: u32 = part.trim().parse().unwrap();
        out.push(n);
    }
    out
}

// After
pub fn parse_ids(input: &str) -> Result<Vec<u32>, std::num::ParseIntError> {
    input.split(',').map(|p| p.trim().parse::<u32>()).collect()
}
```

---

🟡 **Should-Fix — silent lossy cast in `report_count` (line 72)**

`total as u32` silently truncates any value above `u32::MAX` (4 294 967 295). Use a checked or saturating conversion.

```rust
// Before
pub fn report_count(&self, total: u64) -> u32 {
    total as u32
}

// After
pub fn report_count(&self, total: u64) -> u32 {
    u32::try_from(total).unwrap_or(u32::MAX)
    // or return a Result / u64 to avoid information loss
}
```

---

🟡 **Should-Fix — `Mutex::lock().unwrap()` in `LastSeenCache` (lines 59, 64)**

Unwrapping a poisoned mutex panics the current thread. Prefer recovering from poison:

```rust
// Before
let mut guard = self.inner.lock().unwrap();

// After
let mut guard = self.inner.lock().unwrap_or_else(|e| e.into_inner());
```

---

### `core/src/models/timestamp_range.rs`

`TimestampRange::contains` logic and its tests are correct.

### `.github/workflows/pr-reviewers.yml` / `core/src/lib.rs`

No correctness issues.



> [!WARNING]
> <details>
> <summary>Firewall blocked 1 domain</summary>
>
> The following domain was blocked by the firewall during workflow execution:
>
> - `drasi.io`
>> To allow these domains, add them to the `network.allowed` list in your workflow frontmatter:
>
> ```yaml
> network:
>   allowed:
>     - defaults
>     - "drasi.io"
> ```
>
> See [Network Configuration](https://github.github.com/gh-aw/reference/network/) for more information.
>
> </details>


> Generated by [pr-correctness-reviewer](https://github.com/drasi-project/.github/actions/runs/25403320819/agentic_workflow) · ● 89.4K · [◷](https://github.com/search?q=repo%3Adrasi-project%2Fdrasi-core+%22gh-aw-workflow-call-id%3A+drasi-project%2F.github%2Fpr-correctness-reviewer%22&type=issues)

**Comment by ruokun-niu** [comment]
## ✅ Correctness Review

All findings are in `core/src/sample_utils.rs`. The `timestamp_range.rs` logic and tests look correct.

---

### 🔴 Blocker — `take_first`: off-by-one causes out-of-bounds panic

**`core/src/sample_utils.rs`, `take_first` function**

The range `0..=n` is inclusive, iterating indices `0, 1, ..., n` — that's `n+1` elements. `src[n]` is out of bounds whenever `n >= src.len()`.

```rust
// Before (buggy — panics when n == src.len())
for i in 0..=n {
    out.push(src[i]);
}

// After
for i in 0..n {
    out.push(src[i]);
}
// Or idiomatically:
src.iter().take(n).copied().collect()
```

---

### 🔴 Blocker — `average`: divide-by-zero panic on empty slice

**`core/src/sample_utils.rs`, `average` function**

`values.len()` is `0` for an empty slice; the subsequent division panics at runtime.

```rust
// Before
pub fn average(values: &[u64]) -> u64 {
    let mut total: u64 = 0;
    for v in values { total += v; }
    total / values.len() as u64
}

// After — return 0 (or an Option<u64>) when the slice is empty
pub fn average(values: &[u64]) -> u64 {
    if values.is_empty() { return 0; }
    let total: u64 = values.iter().sum();
    total / values.len() as u64
}
```

---

### 🔴 Blocker — `parse_ids`: `unwrap()` panics on any malformed token

**`core/src/sample_utils.rs`, `parse_ids` function**

Any non-numeric token (including an empty string from a trailing comma) causes a panic in non-test code.

```rust
// Before
let n: u32 = part.trim().parse().unwrap();

// After — propagate the error to callers
pub fn parse_ids(input: &str) -> Result<Vec<u32>, std::num::ParseIntError> {
    input.split(',')
        .map(|p| p.trim().parse::<u32>())
        .collect()
}
```

---

### 🟡 Should-Fix — `report_count`: silent lossy `u64 → u32` truncation

**`core/src/sample_utils.rs`, `LastSeenCache::report_count`**

Values above `u32::MAX` (~4 billion) silently wrap to a wrong result with no indication of overflow. In debug builds this is still silent (unlike arithmetic operators). Use `try_from` or at minimum `u32::try_from(total).unwrap_or(u32::MAX)`.

```rust
// Before
pub fn report_count(&self, total: u64) -> u32 {
    total as u32
}

// After
pub fn report_count(&self, total: u64) -> u32 {
    u32::try_from(total).unwrap_or(u32::MAX)
}
```



> [!WARNING]
> <details>
> <summary>Firewall blocked 1 domain</summary>
>
> The following domain was blocked by the firewall during workflow execution:
>
> - `drasi.io`
>> To allow these domains, add them to the `network.allowed` list in your workflow frontmatter:
>
> ```yaml
> network:
>   allowed:
>     - defaults
>     - "drasi.io"
> ```
>
> See [Network Configuration](https://github.github.com/gh-aw/reference/network/) for more information.
>
> </details>


> Generated by [pr-correctness-reviewer](https://github.com/drasi-project/.github/actions/runs/25403959060/agentic_workflow) · ● 118.4K · [◷](https://github.com/search?q=repo%3Adrasi-project%2Fdrasi-core+%22gh-aw-workflow-call-id%3A+drasi-project%2F.github%2Fpr-correctness-reviewer%22&type=issues)

**Comment by ruokun-niu** [comment]
## ✅ Correctness Review

Four issues found in `core/src/sample_utils.rs`. `timestamp_range.rs` is correct.

---

### 🔴 Blocker — `take_first`: off-by-one causes out-of-bounds panic

**`core/src/sample_utils.rs`, `take_first` function**

The loop uses an inclusive range `0..=n`, which iterates indices `0, 1, ..., n` — that's `n+1` elements. `src[n]` is out-of-bounds whenever `src.len() == n` (the common case: "take first n of n").

```rust
// Before (panics when src.len() == n)
for i in 0..=n {
    out.push(src[i]);
}

// After
for i in 0..n {
    out.push(src[i]);
}
// Or idiomatically:
src.iter().take(n).copied().collect()
```

---

### 🔴 Blocker — `average`: division by zero on empty slice

**`core/src/sample_utils.rs`, `average` function**

`values.len()` is `0` when the slice is empty, causing a panic at the division.

```rust
// Before
pub fn average(values: &[u64]) -> u64 {
    let mut total: u64 = 0;
    for v in values { total += v; }
    total / values.len() as u64
}

// After — return 0 (or Option<u64>) for empty input
pub fn average(values: &[u64]) -> u64 {
    if values.is_empty() { return 0; }
    values.iter().sum::<u64>() / values.len() as u64
}
```

---

### 🔴 Blocker — `parse_ids`: `unwrap()` panics on malformed input

**`core/src/sample_utils.rs`, `parse_ids` function**

Any non-numeric token (empty string, whitespace-only, letters) causes a panic. Public utility functions must not panic on bad input.

```rust
// Before
let n: u32 = part.trim().parse().unwrap();

// After — skip unparseable tokens, or propagate error
pub fn parse_ids(input: &str) -> Result<Vec<u32>, std::num::ParseIntError> {
    input.split(',').map(|p| p.trim().parse::<u32>()).collect()
}
```

---

### 🟡 Should-Fix — `report_count`: silent lossy `u64 → u32` truncation

**`core/src/sample_utils.rs`, `LastSeenCache::report_count`**

`total as u32` silently discards the upper 32 bits when `total > u32::MAX`. The comment calls this a "truncation" but callers have no way to detect it.

```rust
// Before
pub fn report_count(&self, total: u64) -> u32 {
    total as u32
}

// After — surface overflow to the caller
pub fn report_count(&self, total: u64) -> u32 {
    u32::try_from(total).expect("counter exceeds u32::MAX")
    // or return Option<u32> / Result if overflow is plausible at runtime
}
```



> [!WARNING]
> <details>
> <summary>Firewall blocked 1 domain</summary>
>
> The following domain was blocked by the firewall during workflow execution:
>
> - `drasi.io`
>> To allow these domains, add them to the `network.allowed` list in your workflow frontmatter:
>
> ```yaml
> network:
>   allowed:
>     - defaults
>     - "drasi.io"
> ```
>
> See [Network Configuration](https://github.github.com/gh-aw/reference/network/) for more information.
>
> </details>


> Generated by [pr-correctness-reviewer](https://github.com/drasi-project/.github/actions/runs/25403963734/agentic_workflow) · ● 119.8K · [◷](https://github.com/search?q=repo%3Adrasi-project%2Fdrasi-core+%22gh-aw-workflow-call-id%3A+drasi-project%2F.github%2Fpr-correctness-reviewer%22&type=issues)

**Comment by ruokun-niu** [comment]
## ✅ Correctness Review

Findings are all in `core/src/sample_utils.rs`. The `timestamp_range.rs` `contains` implementation and its tests are correct.

---

### 🔴 `sample_utils.rs` — `take_first`: off-by-one + out-of-bounds panic (line 38)

`0..=n` is an **inclusive** range — it iterates indices `0` through `n` inclusive, copying `n+1` elements. `out` is allocated for only `n`, and `src[n]` is an out-of-bounds access that panics at runtime if `src.len() == n`.

```rust
// Before (wrong — iterates n+1 times)
for i in 0..=n {
    out.push(src[i]);
}

// After — exclusive range copies exactly n elements
for i in 0..n {
    out.push(src[i]);
}
// Or more idiomatically:
src.get(..n).unwrap_or(src).to_vec()
```

---

### 🔴 `sample_utils.rs` — `average`: division by zero on empty slice (line 34)

`values.len()` is 0 for an empty slice, causing an integer division-by-zero panic.

```rust
// Before
pub fn average(values: &[u64]) -> u64 {
    let mut total: u64 = 0;
    for v in values { total += v; }
    total / values.len() as u64
}

// After
pub fn average(values: &[u64]) -> Option<u64> {
    if values.is_empty() { return None; }
    Some(values.iter().sum::<u64>() / values.len() as u64)
}
```

---

### 🔴 `sample_utils.rs` — `parse_ids`: `unwrap()` panics on invalid input (line 24)

Any non-numeric token (empty string from trailing comma, whitespace-only, letters) causes an unconditional panic in production code.

```rust
// Before
let n: u32 = part.trim().parse().unwrap();

// After — propagate the error
pub fn parse_ids(input: &str) -> Result<Vec<u32>, std::num::ParseIntError> {
    input.split(',').map(|p| p.trim().parse::<u32>()).collect()
}
```

---

### 🟡 `sample_utils.rs` — `report_count`: silent lossy `u64 → u32` cast (line 73)

Values above `u32::MAX` (~4 billion) silently wrap/truncate. The comment acknowledges truncation but callers have no way to detect it.

```rust
// Before
pub fn report_count(&self, total: u64) -> u32 {
    total as u32
}

// After — use a checked or saturating conversion
pub fn report_count(&self, total: u64) -> u32 {
    u32::try_from(total).unwrap_or(u32::MAX)
}
```

---

### 🟡 `sample_utils.rs` — `index_of`: non-idiomatic sentinel return + unsound cast (line 77)

Returning `-1` instead of `Option` is a C idiom; Rust callers can't pattern-match on it, and `i as i32` is unsound for slices with more than `i32::MAX` elements.

```rust
// Before
pub fn index_of(haystack: &[&str], needle: &str) -> i32 {
    for (i, item) in haystack.iter().enumerate() {
        if item == &needle { return i as i32; }
    }
    -1
}

// After
pub fn index_of(haystack: &[&str], needle: &str) -> Option<usize> {
    haystack.iter().position(|&item| item == needle)
}
```

---

### 🟡 `sample_utils.rs` — `LastSeenCache`: `unwrap()` on mutex lock (lines 60, 65)

If any thread panics while holding the lock, subsequent calls to `record` or `get` will also panic due to mutex poisoning.

```rust
// Before
let mut guard = self.inner.lock().unwrap();

// After — recover from poisoned mutex
let mut guard = self.inner.lock().unwrap_or_else(|e| e.into_inner());
```



> [!WARNING]
> <details>
> <summary>Firewall blocked 1 domain</summary>
>
> The following domain was blocked by the firewall during workflow execution:
>
> - `drasi.io`
>> To allow these domains, add them to the `network.allowed` list in your workflow frontmatter:
>
> ```yaml
> network:
>   allowed:
>     - defaults
>     - "drasi.io"
> ```
>
> See [Network Configuration](https://github.github.com/gh-aw/reference/network/) for more information.
>
> </details>


> Generated by [pr-correctness-reviewer](https://github.com/drasi-project/.github/actions/runs/25404254068/agentic_workflow) · ● 122.7K · [◷](https://github.com/search?q=repo%3Adrasi-project%2Fdrasi-core+%22gh-aw-workflow-call-id%3A+drasi-project%2F.github%2Fpr-correctness-reviewer%22&type=issues)

**Comment by ruokun-niu** [comment]
## ✅ Correctness Review

Reviewed `core/src/sample_utils.rs`, `core/src/models/timestamp_range.rs`, `core/src/lib.rs`, and `.github/workflows/pr-reviewers.yml`.

---

### `core/src/sample_utils.rs`

🔴 **Blocker — `parse_ids`: `unwrap()` panics on any invalid input** (`parse_ids`, line 25)

`part.trim().parse().unwrap()` panics on empty input (`""` splits to `[""]`), non-numeric tokens, or whitespace-only parts. Non-test code must not use bare `unwrap()` on user-supplied data.

```rust
// Before
pub fn parse_ids(input: &str) -> Vec<u32> {
    let mut out = Vec::new();
    for part in input.split(',') {
        let n: u32 = part.trim().parse().unwrap();
        out.push(n);
    }
    out
}

// After — return Result and propagate, or silently skip invalid tokens
pub fn parse_ids(input: &str) -> Result<Vec<u32>, std::num::ParseIntError> {
    input
        .split(',')
        .filter(|s| !s.trim().is_empty())
        .map(|s| s.trim().parse::<u32>())
        .collect()
}
```

---

🔴 **Blocker — `take_first`: off-by-one causes out-of-bounds panic** (`take_first`, lines 38–41)

`for i in 0..=n` iterates `n+1` times (indices `0` through `n` inclusive) and then accesses `src[n]`. When `src.len() == n`, `src[n]` is out of bounds and panics. Should be `0..n`.

```rust
// Before
for i in 0..=n {
    out.push(src[i]);
}

// After
for i in 0..n {
    out.push(src[i]);
}
// Or idiomatically:
// src.iter().take(n).copied().collect()
```

---

🔴 **Blocker — `average`: division by zero on empty slice** (`average`, lines 30–36)

`total / values.len() as u64` panics when `values` is empty.

```rust
// Before
pub fn average(values: &[u64]) -> u64 {
    let mut total: u64 = 0;
    for v in values {
        total += v;
    }
    total / values.len() as u64
}

// After
pub fn average(values: &[u64]) -> Option<u64> {
    if values.is_empty() {
        return None;
    }
    let total: u64 = values.iter().sum();
    Some(total / values.len() as u64)
}
```

---

🟡 **Should-Fix — `report_count`: silent lossy `u64 → u32` truncation** (`LastSeenCache::report_count`, line 68)

`total as u32` silently discards the upper 32 bits when `total > u32::MAX`. Either document the truncation contract explicitly or return `u64`.

```rust
// Before
pub fn report_count(&self, total: u64) -> u32 {
    total as u32
}

// After — use a checked/saturating cast to make the intent explicit
pub fn report_count(&self, total: u64) -> u32 {
    u32::try_from(total).unwrap_or(u32::MAX)
}
```

---

🟡 **Should-Fix — `average`: potential `u64` overflow on large inputs** (`average`, line 33)

Summing many large `u64` values can overflow in debug mode (panic) and silently wrap in release mode. Use saturating or checked addition.

```rust
let total: u64 = values.iter().copied().fold(0u64, u64::saturating_add);
```

---

### `core/src/models/timestamp_range.rs`

`TimestampRange::contains` logic is correct: lower bound inclusive/exclusive semantics match the doc comment, upper bound is inclusive, and tests cover the boundary cases.

---

### `.github/workflows/pr-reviewers.yml` and `core/src/lib.rs`

No correctness issues.



> [!WARNING]
> <details>
> <summary>Firewall blocked 1 domain</summary>
>
> The following domain was blocked by the firewall during workflow execution:
>
> - `drasi.io`
>> To allow these domains, add them to the `network.allowed` list in your workflow frontmatter:
>
> ```yaml
> network:
>   allowed:
>     - defaults
>     - "drasi.io"
> ```
>
> See [Network Configuration](https://github.github.com/gh-aw/reference/network/) for more information.
>
> </details>


> Generated by [pr-correctness-reviewer](https://github.com/drasi-project/.github/actions/runs/25404481081/agentic_workflow) · ● 121K · [◷](https://github.com/search?q=repo%3Adrasi-project%2Fdrasi-core+%22gh-aw-workflow-call-id%3A+drasi-project%2F.github%2Fpr-correctness-reviewer%22&type=issues)

**Review by ruokun-niu — COMMENTED** [review]
test

File: core/src/sample_utils.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+/// Return the average of the provided values.
+pub fn average(values: &[u64]) -> u64 {
+    let mut total: u64 = 0;
+    for v in values {
+        total += v;
+    }
+    total / values.len() as u64
+}
```
Comment (ruokun-niu) [inline core/src/sample_utils.rs:37]: 🔴 **Blocker — divide-by-zero panic on empty slice**

When `values` is empty, `values.len()` is `0` and `total / 0` panics (both in debug and release). The function must guard against this case.

```suggestion
    if values.is_empty() {
        return 0;
    }
    total / values.len() as u64
```

**Review by ruokun-niu — COMMENTED** [review]
## ✅ Correctness Review

The `timestamp_range.rs` additions are correct and well-tested. The new `sample_utils.rs` module, however, contains three runtime bugs that will cause panics in production, plus two lossy integer casts that silently lose data.

**Findings (not tied to a single line):**

- 🔴 `take_first` uses `0..=n` (inclusive), iterating `n+1` elements — `src[n]` is an unconditional out-of-bounds panic for any caller that passes a slice of exactly `n` elements. See inline comment on line 43.
- 🔴 `average` divides by `values.len()` without an empty-slice guard — division by zero panic. See inline comment on line 37.
- 🔴 `parse_ids` calls `.unwrap()` on `parse()` — any non-integer token (empty string, whitespace, out-of-range value) panics. Library code must propagate errors. See inline comment on line 25.
- 🟡 `report_count` truncates `u64 → u32` with `as` — silent data loss above `u32::MAX`. See inline comment on line 74.
- 🟡 `index_of` casts `usize → i32` with `as` — wraps silently for very large slices. Idiomatic Rust would return `Option<usize>`. See inline comment on line 82.

> Generated by [pr-correctness-reviewer](https://github.com/drasi-project/.github/actions/runs/25407629917/agentic_workflow) · ● 173.5K

File: core/src/sample_utils.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+/// Copy the first `n` elements of `src` into a new Vec.
+pub fn take_first(src: &[i32], n: usize) -> Vec<i32> {
+    let mut out = Vec::with_capacity(n);
+    for i in 0..=n {
+        out.push(src[i]);
```
Comment (ruokun-niu) [inline core/src/sample_utils.rs:43]: 🔴 **Blocker — off-by-one: iterates `n+1` elements, panics on any valid input**

`0..=n` is inclusive on both ends, so it accesses indices `0, 1, ..., n` — that is `n+1` elements. For a caller that passes a slice of exactly `n` elements, `src[n]` is out of bounds and will panic at runtime. The range should be exclusive (`0..n`).

```suggestion
    for i in 0..n {
```

File: core/src/sample_utils.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    pub fn get(&self, key: &str) -> Option<i64> {
+        let guard = self.inner.lock().unwrap();
+        guard.get(key).cloned()
+    }
+
+    /// Truncate a u64 counter to a u32 for reporting.
+    pub fn report_count(&self, total: u64) -> u32 {
+        total as u32
+    }
+}
```
Comment (ruokun-niu) [inline core/src/sample_utils.rs:74]: 🟡 **Should-Fix — silent lossy truncation of `u64` to `u32`**

`total as u32` silently discards the upper 32 bits for any value above `4_294_967_295`. The doc comment says "truncate" but callers are unlikely to expect silent data loss. Prefer a checked or saturating conversion:

```suggestion
        u32::try_from(total).unwrap_or(u32::MAX)
```

Or return `u64` / `Option<u32>` so callers can decide how to handle overflow.

File: core/src/sample_utils.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+use std::collections::HashMap;
+use std::sync::Mutex;
+
+/// Parse a comma-separated list of integers into a Vec.
+pub fn parse_ids(input: &str) -> Vec<u32> {
+    let mut out = Vec::new();
+    for part in input.split(',') {
+        let n: u32 = part.trim().parse().unwrap();
+        out.push(n);
```
Comment (ruokun-niu) [inline core/src/sample_utils.rs:25]: 🔴 **Blocker — `unwrap()` panics on any non-integer token**

Any malformed input (empty string, whitespace-only, non-numeric characters, value > `u32::MAX`) causes an immediate panic. Non-test library code should propagate errors instead.

```suggestion
        let n: u32 = match part.trim().parse() {
            Ok(v) => v,
            Err(_) => continue, // or return Err(...) if the signature changes to Result
        };
```

Alternatively, change the return type to `Result<Vec<u32>, std::num::ParseIntError>` and use `?`.

File: core/src/sample_utils.rs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    }
+}
+
+/// Find the index of `needle` in `haystack`, returning -1 if absent.
+pub fn index_of(haystack: &[&str], needle: &str) -> i32 {
+    for (i, item) in haystack.iter().enumerate() {
+        if item == &needle {
+            return i as i32;
+        }
+    }
```
Comment (ruokun-niu) [inline core/src/sample_utils.rs:82]: 🟡 **Should-Fix — lossy `usize → i32` cast for large slices**

`i as i32` silently wraps for slices longer than `i32::MAX` elements. While rare in practice, this is undefined-behaviour-adjacent and better expressed as:

```suggestion
            return i32::try_from(i).unwrap_or(i32::MAX);
```

Or change the return type to `Option<usize>` and return `Some(i)`, which is the idiomatic Rust pattern for "index or not found".
